xmake/rules/c++/modules/modules_support/gcc.lua

```diff
@@ -0,0 +1,296 @@
+--!A cross-platform build utility based on Lua
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+-- Copyright (C) 2015-present, TBOOX Open Source Group.
+--
+-- @author      ruki
+-- @file        gcc.lua
+--
+
+-- imports
+import("core.tool.compiler")
+import("core.project.project")
+import("core.project.depend")
+import("core.project.config")
+import("utils.progress")
+import("private.action.build.object", {alias = "objectbuilder"})
+import("common")
+
+-- get and create the path of module mapper
+function get_module_mapper()
+    local mapper_file = path.join(config.buildir(), "mapper.txt")
+    if not os.isfile(mapper_file) then
+        io.writefile(mapper_file, "")
+    end
+
+    return mapper_file
+end
+
+-- add a module or header unit into the mapper
+function add_module_to_mapper(file, module, bmi)
+    for line in io.lines(file) do
+        if line:startswith(module .. " ") then
+            return false
+        end
+    end
+
+    local f = io.open(file, "a")
+    f:printf("%s %s\n", module, bmi)
+    f:close()
+
+    return true
+end
+
+-- load parent target with modules files
+function load_parent(target, opt)
+end
+
+-- check C++20 module support
+function check_module_support(target)
+    local compinst = compiler.load("cxx", {target = target})
+
+    local modulesflag = get_modulesflag(target)
+    local modulemapperflag = get_modulemapperflag(target)
+
+    target:add("cxxflags", modulesflag)
+
+    if os.isfile(get_module_mapper()) then
+        os.rm(get_module_mapper())
+    end
+    target:add("cxxflags", modulemapperflag .. get_module_mapper(), {force = true, expand = false})
+end
+
+-- provide toolchain include dir for stl headerunit when p1689 is not supported
+function toolchain_include_directories(target)
+    if is_plat("linux") then
+        return { "/usr/include/**", "/usr/local/include/**" }
+    end
```

Comment:
Perhaps the cross-compilation toolchain should be better supported here.

https://github.com/xmake-io/xmake/blob/4592135696009bd89d2ed50a29cbcc82bcc6efae/xmake/rules/platform/linux/driver/driver_modules.lua#L51-L87

Reply:
is there an equivalent for clang ?

Reply:
Or for now, add assert 

```lua
    if target:is_plat(os.host()) and target:is_arch(os.arch()) then
    else
          assert("cross toolchain is not supported!")
    end
```

Reply:
> is there an equivalent for clang ?

I have not tried, you can try it. maybe work for clang.

Reply:
clang on my macOS
```
$ clang -E -Wp,-v -xc /dev/null
clang -cc1 version 12.0.5 (clang-1205.0.22.11) default target x86_64-apple-darwin20.5.0
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.5/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
# 1 "/dev/null"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 367 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "/dev/null" 2

```